vispy/visuals/image.py

```diff
@@ -310,6 +342,16 @@ def set_data(self, image):
 
         """
         data = np.asarray(image)
+        if np.iscomplexobj(data):
+            self._data_is_complex = True
+            # even though we're only ever using the first and/or second channels,
+            # without the third empty "channel", it doesn't render correctly.
+            # could be a bug in rg32f rendering?
```

Comment:
Is this still relevant? I see you are using `r32f` mode in your GPU texture? Should it be `rg32f`?

Reply:
my understanding of the texture was that even if you're using `rg32f`, the key should still be `r32f` (this is based  on your [comment here](https://github.com/vispy/vispy/blob/df09b3079d71b7d1942481a051b900f145c4cfee/vispy/visuals/_scalable_textures.py#L321) and the num_channels processing [here](https://github.com/vispy/vispy/blob/df09b3079d71b7d1942481a051b900f145c4cfee/vispy/visuals/_scalable_textures.py#L354).  

The very strange thing here is that, if I try to convert the complex data to `np.stack([data.real, data.imag], axis=-1)`, _without_ that extra "empty" third channel (i.e. a true `rg32f` texture), then for some reason, when I sample the data, everything is just the first `r` channel, even if I specifically sample the `g` channel.  I dug into this one for a _long_ time, trying to figure out where that second channel information was being lost, but never found it.  Realized eventually that just making it 3 channel worked.

maybe you know?  I kinda wonder whether `rg32f` actually works at all?  or whether some 1-or-3 channel assumption is being made lower in the code?

Reply:
:thinking: Good point. You're right...at least that makes sense from the way it is described and implemented. I don't see anything obviously wrong in my code that necessarily assumes something about the data being RGB/A that would effect the storage, the `.r` and the `.g` in the shader (only the clims).

Reply:
yeah, me neither! :joy:  your code seems very agnostic... which is why I was so surprised.  but, if I (for instance) manually go in to the shader code and show just the `.r` channel or just the `.g` channel, they were always the same (until I uploaded 3-channel data to the texture).  But all of the data on the CPU was correct (visualized using matplotlib at various places in the `set_data` processing chain)